### DISCUSSION TITLE
How to use os.execv to run raw shell/powershell commend?

### DISCUSSION THREAD

**Discussion by fripSide** [xmake-io/xmake#2181] [Q&A]
I want to merge output files via shell, 
```
cat bin1 > bin2
```
use the following commands, 
```
os.execv("cat bin1 ", {stdout="bin2"}) 
```

The error message is,
> cannot execv(cat build/nrf-sdk-bin.bin ), No such file or directory

Is there a command just execute my input in the shell like python os.system ？

**Comment by waruqi** [comment]
仔细看下文档，os.execv 第二个参数是 arg list  https://xmake.io/#/zh-cn/manual/builtin_modules?id=osexecv

```lua
os.execv("cat", {"bin1"}, {stdout="bin2"}) 
```

另外，有内置的 io.readfile/io.writefile/io.open/io.read/io.write 一堆跨平台接口，干嘛不用

**Reply by fripSide** [reply]
行，只是io重定向一条shell就搞定，每次都用lua写文件把事情搞复杂了，还得去查lua api文档才能写对。不如直接加个命令：os.shell简单粗暴把string直接拿到shell执行，不做任何转义。

powershell和bash很多命令一样的，可以让用户自己负责跨平台。

**Reply by waruqi** [reply]
> powershell和bash很多命令一样的，可以让用户自己负责跨平台。

不一样，cat 在 linux 下独立程序，os.execv 可以执行，win 上像 cmd 里面 echo, cat 啥的属于 cmd 的内置命令，不是独立程序， os.execv 调用的 CreateProcess 是运行不起来的，因为压根没可执行程序，你要通过 cmd /c "cat ..." 去带起来，两端差异大，还得特殊处理，还不如 `io.writefile("bin2", io.readfile("bin1"))` 也就一句话

**Reply by fripSide** [reply]
行，下次遇到这类重定向，直接write file。这里竟然遇到了两次这种问题，第一直觉都是想直接调用shell。

**Reply by fripSide** [reply]
又踩一次坑，io.readfile(gen_bin1, "rb")，竟然无效，还是按照文本文件去读写。rb不生效，非得先open，再read。
```
local file = io.open(merge_bin, "wb")
local bin1 = io.open(gen_bin1, "rb")
print("add "..gen_bin1)
-- file:write(io.readfile(gen_bin1, "rb"))
file:write(bin1:read("*a"))
-- print("add "..gen_bin2)
-- file:write(io.readfile(gen_bin2, "rb"))
file:close()
```

**Reply by waruqi** [reply]
第二个参数不是 mode string

io.readfile(gen_bin1, {encoding = "binary"})
